geodataRetriever.py

```diff
@@ -58,72 +77,110 @@ def SendEmail(subject, message, TO = toEmails, FROM = fromEmail, password = from
-    SendEmail('Geodata Retriever failed', 'Data failed to download. ' + downloadURL + '\nSee log at ' + logFilePath + '\n'  + traceback.format_exc())
-    sys.exit()
-
-print("unzipping...")
-try:
-    with zipfile.ZipFile(saveLocation, 'r') as f:
-        f.extractall(saveFolder)
-except Exception:
-    logging.exception("Unzip failed")
-    SendEmail('Geodata Retriever failed', 'Unzip failed.\nSee log at ' + logFilePath + '\n' + traceback.format_exc())
-    sys.exit()
-
-# Get field names 
-shapefileFields = [f.name for f in arcpy.ListFields(shapefile) if f.type != 'OID' and f.type != 'Geometry']
-fcFields = [f.name for f in arcpy.ListFields(dataName) if f.type != 'OID' and f.type != 'Geometry']
-
-# Remove FID, OBJECTID
-fields = []
-for f in shapefileFields:
-    if f != 'FID' and f != 'OBJECTID':
-        fields.append(f)
-
-# Check that all fields in shapefile are in target feature class
-if not set(fields).issubset(set(fcFields)):
-    logging.error("Fields in shapefile do not match fields in target feature class.")
-    SendEmail('Geodata Retriever failed', "Fields in download do not match fields in target feature class.\nSee log at " + logFilePath)
-    sys.exit()
-
-# Add geometry field
-fields.append('SHAPE@')
-
-# Delete existing data from table if necessary
-print("deleting rows from " + dataName)
-try:
-    arcpy.DeleteRows_management(dataName)
-except Exception:    
-    logging.exception("Delete rows failed")
-    SendEmail('Geodata Retriever failed', 'Delete rows failed.\nSee log at ' + logFilePath + '\n' + traceback.format_exc())
-    sys.exit()
-
-# Load data into Workspace
-print('loading data from ' + shapefile)
-try:
-    with arcpy.da.SearchCursor(shapefile, fields) as sCursor:
-        with arcpy.da.InsertCursor(dataName, fields) as inCursor:
-            for row in sCursor:
-                inCursor.insertRow(row)
-except Exception:    
-    logging.exception("Insert rows failed")
-    SendEmail('Geodata Retriever failed', 'Insert data failed.\nSee log at ' + logFilePath + '\n' + traceback.format_exc())
-    sys.exit()
-
+# For each data item, download data and udpate feature class or table
+for d in data:
+    print('Updating ' + d)
+    # download data
+    if data.get(d).endswith('.zip'):
+        saveLocation = os.path.join(saveFolder, d + '.zip')
+    elif data.get(d).endswith('.csv'):
+        saveLocation = os.path.join(saveFolder, d + '.csv')
+    else:
+        saveLocation = os.path.join(saveFolder, d)
+    success = DownloadAndUnzip(data.get(d), saveLocation, d)
+    if success: 
+        # get shp file names
+        # NOTE: this will not work if multiple shapefiles are downloaded in the same zip
+        if saveLocation.endswith('.zip'):
+            files = []
+            for file in os.listdir(os.path.join(saveFolder, d)):
+                if file.endswith('.shp'):
+                    files.append(os.path.join(saveFolder, d, file))
+            saveLocation = files.pop()
```

Comment:
What happens if there are multiple shapefiles in the zip? Whoever uses this will most likely not see the note in the code. We should either handle multiple shapefiles, or throw. I vote for throwing since handling multiple shapefiles can get hairy.